python-client/giskard/models/base/wrapper.py

```diff
@@ -0,0 +1,258 @@
+import logging
+import pickle
+from abc import ABC, abstractmethod
+from inspect import isfunction, signature
+from pathlib import Path
+from typing import Any, Callable, Iterable, Optional, Union
+
+import cloudpickle
+import numpy as np
+import pandas as pd
+
+from ...core.core import ModelType
+from ...core.validation import configured_validate_arguments
+from ..utils import warn_once
+from .model import BaseModel
+
+logger = logging.getLogger(__name__)
+
+
+class WrapperModel(BaseModel, ABC):
+    """Base class for model wrappers.
+
+    This is subclass of a :class:`BaseModel` that wraps an existing model
+    object and uses it to make inference.
+
+    This class introduces a `data_preprocessing_function` which can be used
+    to preprocess incoming data before it is passed to the underlying model.
+    """
+
+    @configured_validate_arguments
+    def __init__(
+        self,
+        model: Any,
+        model_type: ModelType,
+        data_preprocessing_function: Optional[Callable[[pd.DataFrame], Any]] = None,
+        model_postprocessing_function: Optional[Callable[[Any], Any]] = None,
+        name: Optional[str] = None,
+        feature_names: Optional[Iterable] = None,
+        classification_threshold: Optional[float] = 0.5,
+        classification_labels: Optional[Iterable] = None,
+        batch_size: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        model : Any
+            The model that will be wrapped.
+        model_type : ModelType
+            The type of the model. Must be a value from the :class:`ModelType`
+            enumeration.
+        data_preprocessing_function : Callable[[pd.DataFrame], Any], optional
+            A function that will be applied to incoming data. Default is ``None``.
+        model_postprocessing_function : Callable[[Any], Any], optional
+            A function that will be applied to the model's predictions. Default
+            is ``None``.
+        name : str, optional
+            A name for the wrapper. Default is ``None``.
+        feature_names : Optional[Iterable], optional
+            A list of feature names. Default is ``None``.
+        classification_threshold : float, optional
+            The probability threshold for classification. Default is 0.5.
+        classification_labels : Optional[Iterable], optional
+            A list of classification labels. Default is None.
+        batch_size : Optional[int], optional
+            The batch size to use for inference. Default is ``None``, which
+            means inference will be done on the full dataframe.
+        """
+        super().__init__(model_type, name, feature_names, classification_threshold, classification_labels, **kwargs)
+        self.model = model
+        self.data_preprocessing_function = data_preprocessing_function
+        self.model_postprocessing_function = model_postprocessing_function
+        self.batch_size = batch_size
```

Comment:
If you think `batch_size` is general enough, I'd suggest adding it to `ModelMeta` in `BaseModel`: https://github.com/Giskard-AI/giskard/blob/28e2ca2e42c3354f7508717d7da23d90c6068091/python-client/giskard/models/base/model.py#L126-L134

This is currently not saved anywhere, anytime you add new attributes to a daughter class, you'd need to save it so that, upon loading (which you also need to override), you feed it to the constructor, see what we did in this class for `data_preprocessing_function` and `model_postprocessing_function`: https://github.com/Giskard-AI/giskard/blob/28e2ca2e42c3354f7508717d7da23d90c6068091/python-client/giskard/models/base/wrapper.py#L217-L226

or also what we did here: https://github.com/Giskard-AI/giskard/blob/28e2ca2e42c3354f7508717d7da23d90c6068091/python-client/giskard/models/pytorch.py#L200-L229

Reply:
I’m not sure I want to save the `batch_size`, because it depends on hardware (and potentially dataset). So when you run it on the ml worker situation might be different.

Actually the same would hold for the torch device…

Reply:
@andreybavt what is your take on this? Should we save this kind of attributes like `device` or `batch_size` which depend on the hardware performing the inference?

My opinion is that they are bound to the local environment. If we envision a use case where models can be downloaded and executed by workers which are not necessarily on the same hardware, this could lead to failures.

Example:
I trained and uploaded my pytorch model on my GPU machine, I set `device="cuda"` and large `batch_size`.
Now I run my worker on a smaller machine which has less GPU memory. When I try to run `model.predict`, I get an out of memory error.
Similar case: I start my worker on my laptop which has no GPU, as soon as it tries to load the model from the giskard server it fails because `device="cuda"` is not a valid device.

Whether this kind of situation would occur depends on the expected usage of the worker.

Reply:
I agree that saving and loading `device` and `batch_size` can be risky when running a worker on a different machine. Our load method already allows passing `kwargs` so these parameters can be specified at the load time.
I think we should either not save them at all or save and load (like we already do for pytorch) but prefer the values from `kwargs` if they're provided.

Reply:
I agree that `batch_size` (post-inference) and `device` are not model properties, but the moment a user wraps their model with a specific `batch_size` and `device`, they (I think) expect the default outcome of this, is that these are the parameters that will be user throughout giskard -- either in the scan (pre-saving) or on the server (post-loading).

So I prefer 

> save and load (like we already do for pytorch) but prefer the values from kwargs if they're provided 

I find it a bit strange that a user wraps his model with `device="cuda"` but suddenly when they upload it to the server and run it, it would run on the default `device="cpu"` (the download/load on the backend happens implicitly, so a user can't specify the `device` when he runs the model on the server after upload).